cli/beaconClient.go

```diff
@@ -0,0 +1,154 @@
+package main
+
+import (
+	"context"
+	"time"
+
+	eth2client "github.com/attestantio/go-eth2-client"
+	"github.com/attestantio/go-eth2-client/api"
+	v1 "github.com/attestantio/go-eth2-client/api/v1"
+	"github.com/attestantio/go-eth2-client/http"
+	"github.com/attestantio/go-eth2-client/spec"
+	"github.com/attestantio/go-eth2-client/spec/phase0"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+)
+
+const slotsPerEpoch = 32
+
+var (
+	ErrNoEigenPod               = errors.New("no eigenpod associated with owner")
+	ErrBeaconClientNotSupported = errors.New("could not instantiate beacon chain client")
+	ErrValidatorNotFound        = errors.New("validator not found")
+)
+
+type BeaconClient interface {
+	GetLatestEpoch(ctx context.Context) (phase0.Epoch, error)
+	GetLatestSlot(ctx context.Context) (phase0.Slot, error)
+	GetBeaconHeader(ctx context.Context, blockId string) (*v1.BeaconBlockHeader, error)
+	GetSignedBeaconBlock(ctx context.Context, blockId string) (*spec.VersionedSignedBeaconBlock, error)
+	GetBeaconState(ctx context.Context, stateId string) (*spec.VersionedBeaconState, error)
+	GetValidator(ctx context.Context, stateID string, validatorIndex phase0.ValidatorIndex) (*v1.Validator, error)
+	GetChainGenesisTime(ctx context.Context) (time.Time, error)
+}
+
+type beaconClient struct {
+	eth2client eth2client.Service
+}
+
+func NewBeaconClient(endpoint string) (BeaconClient, context.CancelFunc, error) {
+	beaconClient := beaconClient{}
+	ctx, cancel := context.WithCancel(context.Background())
+
+	client, err := http.New(ctx,
+		// WithAddress supplies the address of the beacon node, as a URL.
+		http.WithAddress(endpoint),
+		http.WithLogLevel(zerolog.WarnLevel),
+		http.WithTimeout(300*time.Second),
+	)
+	if err != nil {
+		return nil, cancel, err
+	}
+	log.Info().Msgf("Connected to %s\n", client.Name())
+
+	beaconClient.eth2client = client
+	return &beaconClient, cancel, nil
+}
+
+func (b *beaconClient) GetBeaconHeader(ctx context.Context, blockId string) (*v1.BeaconBlockHeader, error) {
+	if provider, isProvider := b.eth2client.(eth2client.BeaconBlockHeadersProvider); isProvider {
+		opts := &api.BeaconBlockHeaderOpts{Block: blockId}
+		response, err := provider.BeaconBlockHeader(ctx, opts)
+		if err != nil {
+			return nil, err
+		}
+		return response.Data, nil
+	}
+
+	return nil, ErrBeaconClientNotSupported
+}
+
+func (b *beaconClient) GetSignedBeaconBlock(ctx context.Context, blockId string) (*spec.VersionedSignedBeaconBlock, error) {
+	if provider, ok := b.eth2client.(eth2client.SignedBeaconBlockProvider); ok {
+		opts := &api.SignedBeaconBlockOpts{Block: blockId}
+		response, err := provider.SignedBeaconBlock(ctx, opts)
+		if err != nil {
+			return nil, err
+		}
+		return response.Data, nil
+	}
+
+	return nil, ErrBeaconClientNotSupported
+}
+
+func (b *beaconClient) GetValidator(ctx context.Context, stateID string, validatorIndex phase0.ValidatorIndex) (*v1.Validator, error) {
```

Comment:
u never use a lot of these funcs, can we take out the unused ones?

Reply:
yeah this file was lifted from the other backend repo, gonna kill off the unused functions 👍